### PR TITLE
Add limit_tokens plugin

### PR DESCRIPTION
**Pull request by MSzabi**
## Type of Change

- **Markdown-It Plugins:** Limit_tokens plugin

## What issue does this relate to?
N/A

### What should this PR do?
Introduces a new plugin with syntax for filtering/transforming tokens in the token stream.

### What are the acceptance criteria?
When the plugin is enabled, only the configured tokens are rendered (or transformed).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1178,6 +1178,30 @@ require('@digitalocean/do-markdownit/vendor/prismjs/plugins/copy-to-clipboard/pr
...
+</details>
+
+```js
+const Prism = require('@digitalocean/do-markdownit/vendor/prismjs');
+
+require('@digitalocean/do-markdownit/vendor/prismjs/plugins/toolbar/prism-toolbar')(Prism);
+require('@digitalocean/do-markdownit/vendor/prismjs/plugins/copy-to-clipboard/prism-copy-to-clipboard')(Prism);
+
+Prism.highlightAll();
+```
```
Comment (MattIPv4) [inline README.md:1203]: This seems like the wrong place to put this, as the Prism Keep HTML plugin is now a subheading of this.

Comment (MattIPv4) [inline README.md:1203]: This seems rather unrelated?

File: modifiers/limit_tokens.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        // Check if current token is in the transformation list
+        const transformOptions = optsObj.transformTokens[token.type];
+        if (transformOptions) {
+            newToken = new token.constructor(
+                transformOptions.type,
+                transformOptions.htmlTag,
+                token.nesting,
+            );
+            newToken.content = token.content;
+        }
```
Comment (MattIPv4) [inline modifiers/limit_tokens.js:62]: I think instead of having a hard mapping like this, `optsObj.transformTokens` should be an `Object<string, function(token): token>`, so that the transformer can do whatever it wants to the token. Some tokens have children etc. that might also need to be adjusted while in the context of the parent

File: modifiers/limit_tokens.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+/**
+ * @module modifiers/limit_tokens
+ */
+
+const safeObject = require('../util/safe_object');
+
+/**
+ * @typedef {Object} LimitTokensOptions
+ * @property {string[]} [allowedTokens] List of MD tokens allowed to render.
+ * @property {Object} [transformTokens] List of MD tokens to transform into something else.
```
Comment (MattIPv4) [inline modifiers/limit_tokens.js:28]: We should properly type this object, so that users know how to interact with it

File: README.md [hunk 4]
```diff
@@ -1178,6 +1178,30 @@ require('@digitalocean/do-markdownit/vendor/prismjs/plugins/copy-to-clipboard/pr
...
+
+<details>
+<summary>Filters and transforms tokens in the token stream.</summary>
+
+**Options:**
+
+Pass options for this plugin as the `limit_tokens` property of the `do-markdownit` plugin options.
+Set this property to `false` to disable this plugin.
+
+- `allowedTokens` (`array`): A list Markdown tokens that should render
```
Comment (davidcarlsonberg) [inline README.md:1191]:
```suggestion
- `allowedTokens` (`array`): A list of Markdown tokens that should render
```

File: README.md [hunk 5]
```diff
@@ -1178,6 +1178,30 @@ require('@digitalocean/do-markdownit/vendor/prismjs/plugins/copy-to-clipboard/pr
...
+- `transformTokens` (`object`): An object where the keys are Markdown tokens that should be transformed.
+The transformation is done base on the value which is an `object` with two keys: `type` and `htmlTag`.
```
Comment (davidcarlsonberg) [inline README.md:1193]:
```suggestion
The transformation is done based on the value which is an `object` with two keys: `type` and `htmlTag`.
```

File: modifiers/limit_tokens.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+ * @type {import('markdown-it').PluginWithOptions<LimitTokensOptions>}
+ */
+module.exports = (md, options) => {
+    // Get the correct options
+    const optsObj = safeObject(options);
+
+    /**
+     * Helper function that does the actual filtering/transforming of tokens.
+     *
+     * @param {import('markdown-it/lib/token').Token[]} tokens The list of tokens to be filtered/transformed
```
Comment (davidcarlsonberg) [inline modifiers/limit_tokens.js:43]:
```suggestion
     * @param {import('markdown-it/lib/token').Token[]} tokens The list of tokens to be filtered/transformed.
```

File: modifiers/limit_tokens.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+     * @returns {import('markdown-it/lib/token').Token[]}
+     */
+    const doFiltering = tokens => tokens.reduce((newStream, token) => {
+        let newToken;
+        // Check if current token is in the allowed list
```
Comment (davidcarlsonberg) [inline modifiers/limit_tokens.js:48]:
```suggestion
        // Check if the current token is in the allowed list
```

File: modifiers/limit_tokens.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        if (optsObj.allowedTokens.includes(token.type)) {
+            newToken = token;
+        }
+
+        // Check if current token is in the transformation list
```
Comment (davidcarlsonberg) [inline modifiers/limit_tokens.js:53]:
```suggestion
        // Check if the current token is in the transformation list
```

File: README.md (hunk 1 above)
Comment (MSzabi) [reply README.md:1203]: Can not count to 3... I did not see the difference between the two headings...
Moved the limit_tokens after all the Prism stuff

Comment (MSzabi) [reply README.md:1203]: Looks like an accidental extra duplication.

File: README.md (hunk 4 above)
Comment (MSzabi) [reply README.md:1191]: Updated.

File: modifiers/limit_tokens.js (hunk 6 above)
Comment (MSzabi) [reply modifiers/limit_tokens.js:43]: Updated.

File: modifiers/limit_tokens.js (hunk 8 above)
Comment (MSzabi) [reply modifiers/limit_tokens.js:53]: Updated.

File: modifiers/limit_tokens.js (hunk 2 above)
Comment (MSzabi) [reply modifiers/limit_tokens.js:62]: Changed the functionality as suggested.

File: modifiers/limit_tokens.js (hunk 3 above)
Comment (MSzabi) [reply modifiers/limit_tokens.js:28]: Updated the type defs.

File: README.md [hunk 9]
```diff
@@ -1197,6 +1197,22 @@ require('@digitalocean/do-markdownit/util/prism_keep_html')(Prism);
...
+<summary>Filters and transforms tokens in the token stream.</summary>
+
+**Options:**
+
+Pass options for this plugin as the `limit_tokens` property of the `do-markdownit` plugin options.
+Set this property to `false` to disable this plugin.
+
+- `allowedTokens` (`array`): A list of Markdown tokens that should render
+- `transformTokens` (`object`): An object where the keys are Markdown tokens that should be transformed.
+  The transformation is done based on the value which is an `function` that expects a Markdown token as a param
```
Comment (MattIPv4) [inline README.md:1212]:
```suggestion
  The transformation is done based on the value which is a `function` that expects a [Markdown token](https://markdown-it.github.io/markdown-it/#Token) as a param
```
Probably worth including a link back to the docs here

File: rules/limit_tokens.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+ */
+module.exports = (md, options) => {
+    // Get the correct options
+    const optsObj = safeObject(options);
+
+    /**
+     * Helper function that does the actual filtering/transforming of tokens.
+     *
+     * @param {token[]} tokens The list of tokens to be filtered/transformed.
+     * @returns {token[]}
```
Comment (MattIPv4) [inline rules/limit_tokens.js:46]:
```suggestion
     * @returns {token[]}
     * @private
```

File: rules/limit_tokens.test.js [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+'use strict';
+
+/**
+ *
```
Comment (MattIPv4) [inline rules/limit_tokens.test.js:20]:
```suggestion
```

File: rules/limit_tokens.test.js [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+ * @typedef {import('markdown-it/lib/token').Token} token
+ */
+
+const fs = require('fs');
+const path = require('path');
+
+const mdEmpty = require('markdown-it')().use(require('..'), {
```
Comment (MattIPv4) [inline rules/limit_tokens.test.js:27]: Instead of importing the whole repo, I would prefer our tests are isolated to this plugin specifically, please

**Review by MattIPv4 — CHANGES_REQUESTED** [review]

File: README.md (hunk 9 above)
Comment (MSzabi) [reply README.md:1212]: Updated.

File: rules/limit_tokens.js (hunk 10 above)
Comment (MSzabi) [reply rules/limit_tokens.js:46]: Updated.

File: rules/limit_tokens.test.js (hunk 12 above)
Comment (MSzabi) [reply rules/limit_tokens.test.js:27]: Updated.

File: rules/limit_tokens.test.js [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    limit_tokens: {
+        allowedTokens: [],
+        transformTokens: {},
+    },
+});
+
+it('does not render any tokens by default', () => {
+    const input = fs.readFileSync(path.join(__dirname, '../fixtures', 'full-input.md'), 'utf8');
```
[the remark this answers is not in the document]
Comment (MSzabi) [reply rules/limit_tokens.test.js:35]: Changed the input to a simple MD text.

File: rules/limit_tokens.test.js [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    expect(mdEmpty.render(input)).toBe('');
+});
+
+const mdSomeAllowed = require('markdown-it')().use(require('..'), {
```
[the remark this answers is not in the document]
Comment (MSzabi) [reply rules/limit_tokens.test.js:39]: Updated.

File: rules/limit_tokens.test.js [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    limit_tokens: {
+        allowedTokens: [ 'caniuse' ],
+        transformTokens: {},
+    },
+});
+
+it('does render allowed tokens', () => {
```
[the remark this answers is not in the document]
Comment (MSzabi) [reply rules/limit_tokens.test.js:37]: Added a new test that ignores not allowed tokens and renders allowed tokens.

File: rules/limit_tokens.js [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+'use strict';
+
+/**
+ * @module rules/limit_tokens
+ *
```
Comment (MattIPv4) [inline rules/limit_tokens.js:21]:
```suggestion
 */
 
/**
```
These need to be separate statements for JSDoc to work

File: rules/limit_tokens.js [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+ * @typedef {import('markdown-it/lib/token').Token} token
```
Comment (MattIPv4) [inline rules/limit_tokens.js:22]:
```suggestion
 * @typedef {import('markdown-it/lib/token').Token} Token
```
Types should be PascalCase

File: rules/limit_tokens.js [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+     */
+    const doFiltering = tokens => tokens.reduce((newStream, token) => {
+        let newToken;
+        // Check if the current token is in the allowed list
+        if (optsObj.allowedTokens.includes(token.type)) {
+            newToken = token;
+        }
+
+        // Check if the current token is in the transformation list
+        if (Object.keys(optsObj.transformTokens).includes(token.type)) {
```
Comment (MattIPv4) [inline rules/limit_tokens.js:57]:
```suggestion
        if (Object.prototype.hasOwnProperty.call(optsObj.transformTokens, token.type)) {
```
We can do a property check here which is far more efficient that iterating over an array of keys

File: rules/limit_tokens.js [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+            newToken = optsObj.transformTokens[token.type](token);
+        }
+
+        if (newToken) {
+            if (token.children) {
+                newToken.children = doFiltering(token.children);
+            }
+
+            newStream.push(newToken);
+        }
```
Comment (MattIPv4) [inline rules/limit_tokens.js:67]:
```suggestion
        // Add the token to the stream, processing any children
        if (newToken) {
            if (token.children) newToken.children = doFiltering(token.children);
            newStream.push(newToken);
        }
```
Cleaning up

**Review by MattIPv4 — CHANGES_REQUESTED** [review]
Apologies, a few more things

File: README.md [hunk 20]
```diff
@@ -1197,6 +1197,22 @@ require('@digitalocean/do-markdownit/util/prism_keep_html')(Prism);
...
+
+<details>
+<summary>Filters and transforms tokens in the token stream.</summary>
+
+**Options:**
+
+Pass options for this plugin as the `limit_tokens` property of the `do-markdownit` plugin options.
+Set this property to `false` to disable this plugin.
+
+- `allowedTokens` (`array`): A list of Markdown tokens that should render
```
[the remark this answers is not in the document]
Comment (MSzabi) [reply README.md:1210]: Changed.

File: README.md [hunk 21]
```diff
@@ -1197,6 +1197,22 @@ require('@digitalocean/do-markdownit/util/prism_keep_html')(Prism);
...
+- `transformTokens` (`object`): An object where the keys are Markdown tokens that should be transformed.
```
[the remark this answers is not in the document]
Comment (MSzabi) [reply README.md:1211]: Changed.

File: README.md [hunk 22]
```diff
@@ -1197,6 +1197,22 @@ require('@digitalocean/do-markdownit/util/prism_keep_html')(Prism);
...
+  The transformation is done based on the value which is a `function` that expects a [Markdown token](https://markdown-it.github.io/markdown-it/#Token) as a param
```
[the remark this answers is not in the document]
Comment (MSzabi) [reply README.md:1212]: Removed.

File: rules/limit_tokens.js (hunk 16 above)
Comment (MSzabi) [reply rules/limit_tokens.js:21]: Fixed.

File: rules/limit_tokens.js (hunk 17 above)
Comment (MSzabi) [reply rules/limit_tokens.js:22]: Fixed.

File: rules/limit_tokens.js [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * @typedef {import('markdown-it/lib/token').Token} token
+ */
+
+const safeObject = require('../util/safe_object');
+
+/**
+ * @typedef {Object} LimitTokensOptions
+ * @property {token.type[]} [allowedTokens] List of MD tokens allowed to render.
+ * @property {Object<string, function(token): token>} [transformTokens] List of MD tokens to transform into something else.
```
[the remark this answers is not in the document]
Comment (MSzabi) [reply rules/limit_tokens.js:30]: Updated.

File: rules/limit_tokens.js [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+ */
+module.exports = (md, options) => {
+    // Get the correct options
+    const optsObj = safeObject(options);
+
+    /**
+     * Helper function that does the actual filtering/transforming of tokens.
+     *
+     * @param {token[]} tokens The list of tokens to be filtered/transformed.
+     * @returns {token[]}
```
[the remark this answers is not in the document]
Comment (MSzabi) [reply rules/limit_tokens.js:46]: Updated.

File: rules/limit_tokens.js [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+     * @private
+     */
+    const doFiltering = tokens => tokens.reduce((newStream, token) => {
+        let newToken;
+        // Check if the current token is in the allowed list
+        if (optsObj.allowedTokens.includes(token.type)) {
+            newToken = token;
+        }
```
[the remark this answers is not in the document]
Comment (MSzabi) [reply rules/limit_tokens.js:54]: Updated.

File: rules/limit_tokens.js (hunk 18 above)
Comment (MSzabi) [reply rules/limit_tokens.js:57]: Updated.

File: rules/limit_tokens.js (hunk 19 above)
Comment (MSzabi) [reply rules/limit_tokens.js:67]: Changed.

**Review by MattIPv4 — APPROVED** [review]
